power_grid_model_c/power_grid_model_cpp/include/power_grid_model_cpp/dataset.hpp

```diff
@@ -171,6 +182,21 @@ class DatasetConst {
     detail::UniquePtr<RawConstDataset, &PGM_destroy_dataset_const> dataset_;
     DatasetInfo info_;
 };
+
+struct OwningMemory {
+    std::vector<Buffer> buffers;
+    std::vector<std::vector<Idx>> indptrs;
+};
+
+struct OwningDatasetMutable {
+    DatasetMutable dataset;
+    OwningMemory storage{};
+};
+
+struct OwningDatasetConst {
+    DatasetConst dataset;
+    OwningMemory storage{};
+};
```

Comment:
i initially tried using one single mutable struct but it gave a multitude of issues. did not want to dive into it while blocking the rest of the work in this PR, so I decided to leave it be for now. since the cpp API wrapper is still experimental, it should be fine.

while people are reviewing, i will spend a small amount of effort to dive into this as well and then we'll see what happens

Reply:
done